src/sage/topology/moment_angle_complex.py

```diff
@@ -814,3 +819,712 @@ def has_trivial_lowest_deg_massey_product(self):
+        sage: x = H.basis(5)[5, 0]
+        sage: y = H.basis(5)[5, 1]
+        sage: x * y
+        0
+    """
+    def __init__(self, base_ring, moment_angle_complex):
+        """
+        Initialize ``self``.
+
+        TESTS::
+
+            sage: S2 = simplicial_complexes.Sphere(2)
+            sage: Z = MomentAngleComplex(S2)
+            sage: H = Z.cohomology_ring(GF(5))
+            sage: TestSuite(H).run()
+            sage: H = Z.cohomology_ring(ZZ)
+            sage: TestSuite(H).run()
+        """
+        self._complex = moment_angle_complex
+        self._base_ring = base_ring
+
+        vertices = moment_angle_complex._simplicial_complex.vertices()
+        n = len(vertices)
+        self._graded_indices = {}
+
+        # Will be used for storing information about the subcomplexes
+        # from which we compute the cohomology
+        self._gens = {}
+        indices = []
+        for deg in range(moment_angle_complex.dimension() + 1):
+            num_of_gens = 0
+            self._gens[deg] = []
+            for i in range(n+1):
+                for x in combinations(vertices, i):
+                    S = moment_angle_complex._simplicial_complex.generated_subcomplex(x, is_mutable=False)
+                    # Because of the empty combination
+                    if len(S.vertices()) > 0 and isinstance(S.cohomology(deg-i-1, base_ring, generators=True), list):
+                        chmlgy = S.cohomology(deg-i-1, base_ring, generators=True)
+                        for y in chmlgy:
+                            self._gens[deg].append((set(x), deg-i-1, y))
+                        num_of_gens += len(chmlgy)
+                    elif len(S.vertices()) == 0 and deg == 0:
+                        num_of_gens = 1
+
+            indices.extend([(deg, k) for k in range(num_of_gens)])
+            self._graded_indices[deg] = range(num_of_gens)
+
+        cat = Algebras(base_ring).WithBasis().Graded().FiniteDimensional()
+        CombinatorialFreeModule.__init__(self, base_ring, indices, category=cat)
+
+    def basis(self, d=None):
+        """
+        Return (the degree ``d`` homogeneous component of) the basis
+        of ``self``.
+
+        INPUT:
+
+        - ``d`` -- (optional) the degree
+
+        EXAMPLES::
+
+            sage: S2 = simplicial_complexes.Sphere(2)
+            sage: Z = MomentAngleComplex(S2)
+            sage: H = Z.cohomology_ring()
+            sage: H.basis()
+            Finite family {(0, 0): h^{0,0}, (7, 0): h^{7,0}}
+            sage: H.basis(5)
+            Finite family {}
+            sage: H.basis(7)
+            Finite family {(7, 0): h^{7,0}}
+            sage: H.basis()[7, 0]
+            h^{7,0}
+            sage: H.basis(8)
+            Finite family {}
+        """
+        if d is None:
+            return Family(self._indices, self.monomial)
+        else:
+            indices = [(d, i) for i in self._graded_indices.get(d, [])]
+            return Family(indices, self.monomial)
```

Comment:
```suggestion
        if d is None:
            return Family(self._indices, self.monomial)
        indices = [(d, i) for i in self._graded_indices.get(d, [])]
        return Family(indices, self.monomial)
```
I find this slightly easier to parse.